pkg/watch/watch.go

```diff
@@ -178,34 +187,47 @@ func (o *WatchClient) WatchAndPush(out io.Writer, parameters WatchParameters) er
 	// mutex as they are shared between goroutines to communicate
 	// sync state/events.
 	var (
-		changeLock   sync.Mutex
-		dirty        bool
-		lastChange   time.Time
-		watchError   error
-		deletedPaths []string
-		changedFiles []string
+		watchError                     error
+		deletedPaths                   []string
+		changedFiles                   []string
+		hasFirstSuccessfulPushOccurred bool
+		wg                             sync.WaitGroup
+		showWaitingMessage             bool
 	)
 
+	//ticker := time.NewTicker(1 * time.Second)
 	watcher, err := fsnotify.NewWatcher()
 	if err != nil {
 		return fmt.Errorf("error setting up filesystem watcher: %v", err)
 	}
 	defer watcher.Close()
 	defer close(parameters.ExtChan)
 
+	// adding watch on the root folder and the sub folders recursively
+	// so directory and the path in addRecursiveWatch() are the same
+	err = addRecursiveWatch(watcher, parameters.Path, parameters.FileIgnores)
+	if err != nil {
+		return fmt.Errorf("error watching source path %s: %v", parameters.Path, err)
+	}
+
+	printInfoMessage(out, parameters.Path)
 	// This goroutine listens for either file change events from fsnotify, fs errors, or a terminate signal
 	// The results are stored in the variables defined in the var( ... ) block above
+	wg.Add(1)
 	go func() {
+		var (
+			timer     *time.Timer
+			lastEvent fsnotify.Event
+		)
+		timer = time.NewTimer(time.Millisecond)
+		<-timer.C
 		for {
 			select {
-			case extMsg := <-parameters.ExtChan:
-				if extMsg {
-					changeLock.Lock()
-					watchError = ErrUserRequestedWatchExit
-					changeLock.Unlock()
-				}
 			case event := <-watcher.Events:
-				changeLock.Lock()
+				lastEvent = event
```

Comment:
You can have several `watcher.Events` before to have `timer.C`, so you need to store your events into an array of events: `lastEvents = append(lastEvents, event)`

Reply:
But wouldn't that cause the same problem because `lastEvents` slice/array would have multiple events triggered for a single change?

Reply:
Probably you could filter it to remove duplicates?

Reply:
Probably. But what would be the benefit compared to adding the logic for filtering? In the interval of 300 ms (the duration of the timer), isn't it safe to assume there won't be too many events that we would need to filter them?

PS: I might be horribly misunderstanding `timer.Timer`, so correct me if I am wrong.

Reply:
Makes sense. I'll append events and keep the unique ones. 👍🏾